### PR TITLE
Sapling note encryption

### PR DESCRIPTION
**Pull request by str4d**
Implements full trial decryption (per the protocol spec) as well as compact trial decryption (per ZIP 307).

### CODE REVIEW

File: zcash_primitives/src/keys.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    jubjub::{edwards, FixedGenerators, JubjubEngine, JubjubParams, ToUniform, Unknown},
+    primitives::{ProofGenerationKey, ViewingKey},
+};
+use std::io::{self, Read, Write};
+
+pub const PRF_EXPAND_PERSONALIZATION: &'static [u8; 16] = b"Zcash_ExpandSeed";
+
+/// PRF^expand(sk, t) := BLAKE2b-512("Zcash_ExpandSeed", sk || t)
+pub fn prf_expand(sk: &[u8], t: &[u8]) -> Blake2bResult {
+    prf_expand_vec(sk, &vec![t])
```
Comment (ebfull) [inline zcash_primitives/src/keys.rs]: Only need to make a slice not a vector.

File: zcash_primitives/src/note_encryption.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    for i in 0..buffer.len() {
+        buffer[i] = rng.gen();
+    }
+
+    // reduce to uniform value
+    Fs::to_uniform(&buffer[..])
+}
+
+fn sapling_ka_agree(esk: &Fs, pk_d: &edwards::Point<Bls12, PrimeOrder>) -> Vec<u8> {
+    let ka = pk_d
```
Comment (ebfull) [inline zcash_primitives/src/note_encryption.rs]: This should more closely mirror the previous code

```rust

    // Deserialize sk
    let sk = match Fs::from_repr(read_fs(&(unsafe { &*sk })[..])) {
        Ok(p) => p,
        Err(_) => return false,
    };

    // Multiply by 8
    let p = p.mul_by_cofactor(&JUBJUB);

    // Multiply by sk
    let p = p.mul(sk, &JUBJUB);

    // Produce result
    let result = unsafe { &mut *result };
    p.write(&mut result[..]).expect("length is not 32 bytes");
```

File: zcash_primitives/src/note_encryption.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    pub fn epk(&self) -> &edwards::Point<Bls12, PrimeOrder> {
+        &self.epk
+    }
+
+    pub fn encrypt_note_plaintext(&self) -> [u8; 580] {
+        let shared_secret = sapling_ka_agree(&self.esk, &self.to.pk_d);
+        let key = kdf_sapling(&shared_secret, &self.epk);
+
+        let nonce = [0u8; 12];
```
Comment (ebfull) [inline zcash_primitives/src/note_encryption.rs]: Inline this.

File: zcash_primitives/src/note_encryption.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+        let mut input = Vec::with_capacity(564);
+        input.push(1);
+        input.extend_from_slice(&self.to.diversifier.0);
+        (&mut input)
+            .write_u64::<LittleEndian>(self.note.value)
+            .unwrap();
+        self.note.r.into_repr().write_le(&mut input).unwrap();
+        input.extend_from_slice(&self.memo.0);
+
```
Comment (ebfull) [inline zcash_primitives/src/note_encryption.rs]: Assert the vector is of length `564`.

Comment (ebfull) [reply zcash_primitives/src/note_encryption.rs]: Might as well also make a constant for `564`.

File: zcash_primitives/src/note_encryption.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        let mut ciphertext = Vec::with_capacity(564);
+        let tag =
+            chacha20_poly1305_aead::encrypt(&key.as_bytes(), &nonce, &[], &input, &mut ciphertext)
+                .unwrap();
+
+        let mut output = [0u8; 580];
```
Comment (ebfull) [inline zcash_primitives/src/note_encryption.rs]: Make a constant for 580.

File: zcash_primitives/src/note_encryption.rs [hunk 6]
```diff
@@ -238,6 +304,20 @@ mod tests {
             let note = to.create_note(tv.v, rcm, &JUBJUB).unwrap();
             assert_eq!(note.cm(&JUBJUB), cmu);
 
+            //
```
Comment (ebfull) [inline zcash_primitives/src/note_encryption.rs]: Add additional tests for failure to decrypt for various reasons:

1. invalid diversifier
2. invalid tag
3. invalid commitment
4. slice length too small?

File: zcash_primitives/src/note_encryption.rs [hunk 7]
```diff
@@ -224,6 +224,88 @@ pub fn try_sapling_note_decryption(
...
+///
+/// Implements section 4.17.3 of the Zcash Protocol Specification.
+pub fn try_sapling_output_recovery(
+    ovk: &OutgoingViewingKey,
+    cv: &edwards::Point<Bls12, Unknown>,
+    cmu: &Fr,
+    epk: &edwards::Point<Bls12, PrimeOrder>,
+    enc_ciphertext: &[u8],
+    out_ciphertext: &[u8],
+) -> Option<(Note<Bls12>, PaymentAddress<Bls12>, Memo)> {
```
Comment (ebfull) [inline zcash_primitives/src/note_encryption.rs]: Assert lengths of input slices.

File: zcash_primitives/src/note_encryption.rs [hunk 8]
```diff
@@ -318,6 +400,15 @@ mod tests {
                 None => panic!("Note decryption failed"),
             }
 
+            match try_sapling_output_recovery(&ovk, &cv, &cmu, &epk, &tv.c_enc, &tv.c_out) {
```
Comment (ebfull) [inline zcash_primitives/src/note_encryption.rs]: Similarly test edge cases like failures of tags or commitments not matching.

**Comment by str4d** [comment]
Addressed @ebfull's comments.

**Comment by str4d** [comment]
Rebased on master to address merge conflicts.

**Comment by str4d** [comment]
In the last month, a new crate has been published that exposes both the ChaCha20Poly1305 IETF construction and the underlying ChaCha20 IETF primitive. I've pushed a commit that switches us from our own fork of an implementation (which would inhibit us publishing crates) to this new crate. The previous test vectors all pass.

Note that this requires raising our minimum Rust version to 1.32. Incidentally, by doing this we should be able to replace our own usage of the `byteorder` crate with [conversion functions on the inbuilt primitives](https://blog.rust-lang.org/2019/01/17/Rust-1.32.0.html#library-stabilizations).

File: zcash_primitives/src/note_encryption.rs [hunk 9]
```diff
@@ -192,36 +222,41 @@ pub fn try_sapling_note_decryption(
...
+    let nonce = [0u8; 12];
+    let mut chacha20 = ChaCha20::new(key.as_bytes(), &nonce);
+    // Skip over Poly1305 keying output
+    chacha20.next();
+
+    let mut plaintext = Vec::with_capacity(52);
+    plaintext.extend_from_slice(&enc_ciphertext[0..52]);
+    let keystream = chacha20.next();
+    for i in 0..52 {
+        plaintext[i] ^= keystream.as_bytes()[i];
```
Comment (Eirik0) [inline zcash_primitives/src/note_encryption.rs]: Nit: Does it make sense to call `.as_bytes()` outside of the loop? I don't know the details of this function, but I feel like the processor might be doing a little more work than is necessary here.
Edit: I see this changes later so never mind.

File: zcash_primitives/src/note_encryption.rs [hunk 10]
```diff
@@ -29,6 +64,58 @@ impl Default for Memo {
...
+
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..]
+    }
+
+    /// Returns:
+    /// - None if the memo is not text
+    /// - Some(Ok(memo)) if the memo contains a valid UTF8 string
+    /// - Some(Err(e)) if the memo contains invalid UTF8
+    pub fn to_utf8(&self) -> Option<Result<String, str::Utf8Error>> {
```
Comment (Eirik0) [inline zcash_primitives/src/note_encryption.rs]: I think having Option<Result<...>> is confusing as Option and Result exist to serve similar purposes. I could see this being justified if we handles Err() and None differently, but in this code it doesn't look like we are doing so.
This function could be changed to:
```
pub fn to_utf8(&self) -> Option<String> {
```
if we changed the line (112):
```
Some(str::from_utf8(data).map(|memo| memo.to_owned()))
```
to
```
match str::from_utf8(data).map(|memo| memo.to_owned()) {
    Ok(utf8) => Some(utf8),
    Err(_) => None,
}
```

File: zcash_primitives/src/note_encryption.rs [hunk 11]
```diff
@@ -20,6 +20,22 @@ use crate::{keys::OutgoingViewingKey, JUBJUB};
 pub const KDF_SAPLING_PERSONALIZATION: &'static [u8; 16] = b"Zcash_SaplingKDF";
 pub const PRF_OCK_PERSONALIZATION: &'static [u8; 16] = b"Zcash_Derive_ock";
 
+const COMPACT_NOTE_SIZE: usize = (
+    1  + // version
+    11 + // diversifier
+    8  + // value
+    32
+    // rcv
```
Comment (Eirik0) [inline zcash_primitives/src/note_encryption.rs]: Nit: It is not clear whether or not 32 and rcv go together. In the three previous values the comments are on the same line. It would be nice to be consistent, but I'm not too worried about it as I could see this being something the formatter did. *shrug*
(similar below)

File: zcash_primitives/src/note_encryption.rs [hunk 12]
```diff
@@ -274,7 +294,7 @@ fn parse_note_plaintext_minus_memo(
     let v = (&plaintext[12..20]).read_u64::<LittleEndian>().ok()?;
 
     let mut rcm = FsRepr::default();
-    rcm.read_le(&plaintext[20..52]).ok()?;
+    rcm.read_le(&plaintext[20..COMPACT_NOTE_SIZE]).ok()?;
```
Comment (Eirik0) [inline zcash_primitives/src/note_encryption.rs]: 20 = version + diversifier + value. Is it worth it to have a constant for that too?

File: zcash_primitives/src/note_encryption.rs (hunk 11 above)
Comment (str4d) [reply zcash_primitives/src/note_encryption.rs]: Yep, this is just how `rustfmt` wants to format it 🤷‍♂

File: zcash_primitives/src/note_encryption.rs (hunk 12 above)
Comment (str4d) [reply zcash_primitives/src/note_encryption.rs]: I considered doing that for both the `20` and the `12`, but it seemed like a lot of textual overhead for boundaries that weren't (for the moment) as important as the `CompactNote` boundary.

File: zcash_primitives/src/note_encryption.rs (hunk 10 above)
Comment (str4d) [reply zcash_primitives/src/note_encryption.rs]: The reason I separated the two cases was so that callers of this API _could_ handle the `Err()` differently from the `None`. See [here](https://github.com/str4d/librustzcash/blob/ae63116f639847e54716fbf47f19589d12038a16/zcash_client_sqlite/src/lib.rs#L544-L551) as an example (where I end up inverting the `Result` and `Option` so that I can combine the error types due to SQL errors vs the memo being invalid).

Given that the error in question here is specifically that the memo field contained invalid UTF-8 when it was supposed to be UTF-8 text, one alternative is to just do as you are, and pretend the memo does not exist if it is invalid. Another option would be to use `String::from_utf8_lossy` instead, and try to return as much useful text as we can. The current case (return the error) gives more information to the caller (who could then choose to manually call `String::from_utf8_lossy` on the memo's bytes if they wanted), but does forgo a slightly simpler interface.

**Comment by daira** [comment]
Please note that merging this also blocks on reviewing the `crypto_api` and `crypto_api_chachapoly` crates. We shouldn't switch to them without verifying that they have identical behaviour on all inputs.

**Comment by str4d** [comment]
@daira we don't yet depend on any Chacha20Poly1305 implementation in the Rust codebase. When I said "switch", I meant that this PR originally added a dependency on (a fork of) one implementation, and it now adds a dependency on a different implementation. I agree that we should still be happy that the dependency is solid for our usage.

File: zcash_primitives/src/note_encryption.rs (hunk 10 above)
Comment (Eirik0) [reply zcash_primitives/src/note_encryption.rs]: I am fine with leaving this as is.

**Comment by daira** [comment]
@defuse, @str4d and I reviewed the `crypto_api_chachapoly` AEAD implementation. I initially thought that there was a problem with this implementation when the AAD (which we don't use) or the plaintext is not a multiple of 16 bytes &mdash; but I was mistaken. The necessary padding is implemented correctly here: https://github.com/KizzyCode/crypto_api_chachapoly/blob/master/src/poly1305.rs#L41-L57

**Comment by str4d** [comment]
Two issues were opened on `crypto_api_chachapoly` as a result of the review, which do not affect this PR:
- https://github.com/KizzyCode/crypto_api_chachapoly/issues/2
  - We only encrypt notes, which are fixed-length.
- https://github.com/KizzyCode/crypto_api_chachapoly/issues/3
  - We do not use the `new_sec_key` function.

File: zcash_primitives/src/note_encryption.rs [hunk 13]
```diff
@@ -0,0 +1,1079 @@
...
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, cmu, epk, mut enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        enc_ciphertext[ENC_CIPHERTEXT_SIZE - 1] ^= 0xff;
+        assert_eq!(
+            try_sapling_note_decryption(&ivk, &epk, &cmu, &enc_ciphertext),
+            None
+        );
+    }
```
Comment (daira) [inline zcash_primitives/src/note_encryption.rs]: There's a missing failure case for all three decryption types: the diversifier may be invalid so that g<sub>d</sub> can't be computed. `random_enc_ciphertext` will never create a ciphertext that triggers this case because it always uses a fixed (all-zeros) diversifier.

File: zcash_primitives/src/note_encryption.rs [hunk 14]
```diff
@@ -0,0 +1,1079 @@
...
+        assert_eq!(
+            try_sapling_compact_note_decryption(
+                &ivk,
+                &epk,
+                &cmu,
+                &enc_ciphertext[..COMPACT_NOTE_SIZE]
+            ),
+            None
+        );
+    }
```
Comment (daira) [inline zcash_primitives/src/note_encryption.rs]: There's a missing failure case for each of normal decryption and for recovery: the diversifier may be invalid so that g<sub>d</sub> can't be computed. `random_enc_ciphertext` will never create a ciphertext that triggers this case because it always uses a fixed (all-zeros) diversifier.

`compact_decryption_with_invalid_diversifier` covers only the case of the diversifier with (little-endian) value 0xFF. I don't know whether this is a valid diversifier or not, but either way it can only be covering one of the two possible cases: where the diversifier itself is invalid, or where it is valid but causes the commitment to be wrong.

File: zcash_primitives/src/note_encryption.rs [hunk 15]
```diff
@@ -0,0 +1,1079 @@
...
+            let mut ne = SaplingNoteEncryption::new(ovk, note, to, Memo(tv.memo));
+            // Swap in the ephemeral keypair from the test vectors
+            ne.esk = esk;
+            ne.epk = epk;
+
+            assert_eq!(&ne.encrypt_note_plaintext()[..], &tv.c_enc[..]);
+            assert_eq!(&ne.encrypt_outgoing_plaintext(&cv, &cmu)[..], &tv.c_out[..]);
+        }
+    }
+}
```
Comment (daira) [inline zcash_primitives/src/note_encryption.rs]: Despite my nitpicking, these are *excellent* tests! :+1:

File: zcash_primitives/src/test_vectors/note_encryption.rs [hunk 16]
```diff
@@ -0,0 +1,2046 @@
...
+    pub k_enc: [u8; 32],
+    pub p_enc: [u8; 564],
+    pub c_enc: [u8; 580],
+    pub ock: [u8; 32],
+    pub op: [u8; 64],
+    pub c_out: [u8; 80],
+}
+
+pub(crate) fn make_test_vectors() -> Vec<TestVector> {
+    // From https://github.com/zcash-hackworks/zcash-test-vectors/blob/master/sapling_note_encryption.py
```
Comment (daira) [inline zcash_primitives/src/test_vectors/note_encryption.rs]: I have not checked that these match the source.

File: zcash_primitives/src/keys.rs [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+            None => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "ak not of prime order",
+                ));
+            }
+        };
+
+        let nk = edwards::Point::<E, Unknown>::read(&mut reader, params)?;
+        let nk = match nk.as_prime_order(params) {
```
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_primitives/src/keys.rs]: `edwards::Point<_, Unknown>::as_prime_order()` is implemented like so:
```rust
    /// Attempts to cast this as a prime order element, failing if it's
    /// not in the prime order subgroup.
    pub fn as_prime_order(&self, params: &E::Params) -> Option<Point<E, PrimeOrder>> {
        if self.mul(E::Fs::char(), params) == Point::zero() {
            Some(convert_subgroup(self))
        } else {
            None
        }
    }
```

So I believe that means `edwards::Point<_, PrimeOrder>` is already of type J<sup>(r)</sup>, since the identity is idempotent. It also means that `ak` is used outside the circuit as type J<sup>(r)</sup> and then constrained further by the circuit to be of type J<sup>(r)*</sup>.

File: zcash_primitives/src/note_encryption.rs [hunk 18]
```diff
@@ -0,0 +1,1079 @@
...
+    // Produce result
+    let mut result = [0; 32];
+    p.write(&mut result[..]).expect("length is not 32 bytes");
+    result
+}
+
+/// Sapling KDF for note encryption.
+///
+/// Implements section 5.4.4.4 of the Zcash Protocol Specification.
+fn kdf_sapling(dhsecret: &[u8], epk: &edwards::Point<Bls12, PrimeOrder>) -> Blake2bResult {
```
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_primitives/src/note_encryption.rs]: `sapling_ka_agree()` was implemented by @gtank to return a byte array, likely because you never use the output of ECDH as a point (or directly at all, since it should always go through a KDF). But it's a fair point (no pun intended) that KA<sup>Sapling</sup>.Agree returns a point, and KDF<sup>Sapling</sup> takes two points.

File: zcash_primitives/src/note_encryption.rs [hunk 19]
```diff
@@ -0,0 +1,1079 @@
...
+                    &mut buf,
+                    OUT_PLAINTEXT_SIZE,
+                    &[],
+                    key.as_bytes(),
+                    &[0u8; 12]
+                )
+                .unwrap(),
+            OUT_CIPHERTEXT_SIZE
+        );
+        buf
```
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_primitives/src/note_encryption.rs]: Opened https://github.com/KizzyCode/crypto_api_chachapoly/issues/4

File: zcash_primitives/src/keys.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_primitives/src/keys.rs]: Ah, I see you looked at `as_prime_order` later on in the review. I'll add an identity check on `ak` after this call.

File: zcash_primitives/src/note_encryption.rs [hunk 20]
```diff
@@ -0,0 +1,1079 @@
...
+    }
+
+    /// Returns:
+    /// - `None` if the memo is not text
+    /// - `Some(Ok(memo))` if the memo contains a valid UTF-8 string
+    /// - `Some(Err(e))` if the memo contains invalid UTF-8
+    pub fn to_utf8(&self) -> Option<Result<String, str::Utf8Error>> {
+        // Check if it is a text or binary memo
+        if self.0[0] < 0xF5 {
+            // Drop trailing zeroes
```
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_primitives/src/note_encryption.rs]: `self.0` is a `[u8; 512]`, not a `str`. We only get a `str` on the next line (because `str` is required to be valid UTF-8 in Rust).

[the remark this answers is not in the document]
Comment (daira) [reply zcash_primitives/src/note_encryption.rs]: Right, but it should be possible to do something like (untested):
```
let s = str::from_utf8(self.0)?;
Some(s.trim_end_matches(char::from(0)));
```
(It's not doing precisely the same thing, but it is equivalent because the trailing NUL characters do not affect whether the string is valid UTF-8.)

File: zcash_primitives/src/note_encryption.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (KizzyCode) [reply zcash_primitives/src/note_encryption.rs]: @daira This is a good point. Currently, the entire ChaCha20 implementation works by creating a key-stream block and XORing this block directly into the target buffer because this is more efficient than creating a key-stream block, XORing this block in-place with a source block and copying the result into the target buffer.
Currently I'm implementing the changes suggested by @str4d (the `seal_to`/`open_to` functions will have their own checks) – I'll keep your point in mind for future development 😊

File: zcash_primitives/src/note_encryption.rs [hunk 21]
```diff
@@ -0,0 +1,1079 @@
...
+    assert_eq!(
+        ChaCha20Ietf::cipher()
+            .decrypt(
+                &mut plaintext,
+                CHACHA20_BLOCK_SIZE + COMPACT_NOTE_SIZE,
+                key.as_bytes(),
+                &[0u8; 12],
+            )
+            .ok()?,
+        CHACHA20_BLOCK_SIZE + COMPACT_NOTE_SIZE
```
[the remark this answers is not in the document]
Comment (KizzyCode) [reply zcash_primitives/src/note_encryption.rs]: Yeah; I guess I could expose this API. As far as I can see, this only requires two additional assertions (to verify key and nonce length). Will look into it later.

File: zcash_primitives/src/note_encryption.rs [hunk 22]
```diff
@@ -0,0 +1,1079 @@
...
+    plaintext: &[u8],
+) -> Option<(Note<Bls12>, PaymentAddress<Bls12>)> {
+    let mut d = [0u8; 11];
+    d.copy_from_slice(&plaintext[1..12]);
+
+    let v = (&plaintext[12..20]).read_u64::<LittleEndian>().ok()?;
+
+    let mut rcm = FsRepr::default();
+    rcm.read_le(&plaintext[20..COMPACT_NOTE_SIZE]).ok()?;
+    let rcm = Fs::from_repr(rcm).ok()?;
```
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_primitives/src/note_encryption.rs]: It's the former, because `sapling_crypto::jubjub` does not use `ff_derive`.

**Comment by str4d** [comment]
Addressed @daira's comments. I also replaced all uses of `Vec` with fixed-length arrays.

File: zcash_primitives/src/note_encryption.rs [hunk 23]
```diff
@@ -0,0 +1,1377 @@
...
+                &enc_ciphertext,
+                &[],
+                key.as_bytes(),
+                &[0u8; 12]
+            )
+            .ok()?,
+        NOTE_PLAINTEXT_SIZE
+    );
+
+    // Check note plaintext version
```
Comment (daira) [inline zcash_primitives/src/note_encryption.rs]: :+1:

**Comment by str4d** [comment]
@daira the comment was added to the top of the file in 23aa869bf423e9778e7eb9e88544e203675d8005.

File: zcash_primitives/src/keys.rs [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+    pub ovk: OutgoingViewingKey,
+}
+
+/// A Sapling full viewing key
+pub struct FullViewingKey<E: JubjubEngine> {
+    pub vk: ViewingKey<E>,
+    pub ovk: OutgoingViewingKey,
+}
+
+impl<E: JubjubEngine> ExpandedSpendingKey<E> {
```
[the remark this answers is not in the document]
Comment (daira) [reply zcash_primitives/src/keys.rs]: ah I see it now

File: zcash_primitives/src/note_encryption.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (KizzyCode) [reply zcash_primitives/src/note_encryption.rs]: Fixed in https://github.com/KizzyCode/crypto_api_chachapoly/commit/4fe55c536dbca9fd62ff31f49e3f188c50667e64
